### DISCUSSION TITLE
マルチプレイ機能を使ったときに相手のアバターが表示されない。

### DISCUSSION THREAD

**Discussion by SK0519** [pixiv/vroid-sdk-developers#188] [Q&A]
下記の手順を実行すると、自分のアバターが表示されますが相手のアバターが表示されず、数秒後にUEが強制終了されます。
https://vroid.notion.site/VRoid-SDK-for-Unreal-Engine-21321b5e31b247e99600f270b3414d4e#322cda5f4e714fc2b74ee005016090f6

また、『Third Person Characterの代わりにVRMを動かす』を実行する際に『コンテンツドロワーの「設定」から「プラグインコンテンツを表示」を有効化』をしているにも関わらず、「BP_ThirdPersonGameMode」が表示されないため、ソロプレイではDefault Pawn Classの欄をBP_VRoidMannequinRetargetCharacterに変更できず、自分のアバターを表示することができません。

①相手のアバターを表示させる方法
②ソロプレイ時に自分のアバターを表示させる方法
以上２点をお聞きしたいです。

UEは5.4を使用しております。
お忙しいところ恐縮ですが何卒宜しくお願い致します。

![スクリーンショット (102)](https://github.com/user-attachments/assets/1e4e77e6-460b-47b5-b189-f95420838ed5)

**Comment by mameoPix** [comment]
ご質問ありがとうございます。
順を追って1つずつ一緒に解決していければと思います。

まず、**BP_ThirdPersonGameModeが表示されない問題**について
こちらはPluginではなく、ThirdPersonTemplateのBlueprintを使用する想定です。
下記のドキュメントに記載があるように、コンテンツ追加からThirdPersonのテンプレートを追加することで表示されるようになります。
[プロジェクトの作成/必要なContentPackについて](https://vroid.notion.site/VRoid-SDK-for-Unreal-Engine-21321b5e31b247e99600f270b3414d4e#91978d36ef0f4297a99f78d17b5e63d4
)
使用するBlueprintのPathは下記になります。
`Content/ThirdPerson/Blueprints/BP_ThirdPersonGameMode
`
> ②ソロプレイ時に自分のアバターを表示させる方法

こちらについては上記のGameModeのDefault Pawn Classを`BP_VRoidMannequinRetargetCharacter`に置き換えることで解決出来るかと思います。


続いて
> ①相手のアバターを表示させる方法

こちらについては、
[Multiplay機能を使用するには](https://vroid.notion.site/VRoid-SDK-for-Unreal-Engine-21321b5e31b247e99600f270b3414d4e#dbd9ec3220d445e3b0e31123a3ea8192)のドキュメントを参考にしてみてください。
こちらの手順で分からない点や、追加のご質問があればコメントにてお願い致します。


最後に、数秒後にUEがクラッシュする というのは、Editorで実行時でしょうか？
それともパッケージ化した際のプロジェクトでしょうか？
もし、よろしければ、下記の[クラッシュした際のLogの取得方法](
https://vroid.notion.site/VRoid-SDK-for-Unreal-Engine-21321b5e31b247e99600f270b3414d4e#5b74774999494e3092a8a10e66086159)を参考に
ログを送っていただければこちらでさらに詳細な調査をお手伝いさせていただきます。

**Reply by SK0519** [reply]
ご返信ありがとうございます。

②について
解決いたしました。ご対応いただきありがとうございました。

①について
ログファイルを添付いたしましたのでご確認のほどよろしくお願いいたします。
[SimpleCollabSystem.log](https://github.com/user-attachments/files/18259992/SimpleCollabSystem.log)

**Reply by mameoPix** [reply]
迅速にログファイルをご提供いただきありがとうございます！
クラッシュ時の内容を確認 / 手元でも不具合が発生することを確認出来ました。

回避策としては、**BP_VRoidMultiplayCharacter**内のSpawn Parameterを更新して
BP_VRoidMultiplayCharacterのPath: `(Plugins/VRoidSdk/Blueprints/Multiplay/BP_VRoidMultiplayCharacter)`
下図の2つのオプションを無効にすることで回避可能かと思います。
・Use Aynsc Load
・Use Fast Converter
![image](https://github.com/user-attachments/assets/a91932e3-0ab0-4a7f-937a-9d8fb352e235)
ただし、これら2つのオプションを無効化することによってVRMのロード時間が長くなってしまう問題があります。
複数のVRMをロードする際や、Loadに時間のかかる比較的重いVRMを使用した際には
ロードの待ち時間が長くなりすぎてクラッシュしてしまう可能性が残っています。
また、まだ手元では確認出来ておりませんが、この不具合はパッケージ化した際には発生せず、Editor実行時でのみ発生する可能性があります。



クラッシュの根本原因としては2つあることが分かりました。
1. 本SDKで実装している高速化したConverterによるVRMの変換時にリソースが競合していることによる問題
2.  非同期読み込みを行う際に1. と同様にSDKで実装している非同期処理でリリースが競合していることによる問題

上記2点については、いずれも解決までに時間がかかってしまいそうです。
大変申し訳ございません。

**Comment by SK0519** [comment]
ご回答ありがとうございます。

頂いた設定で何度か試したところエディター上でマルチプレイをすることに成功しました。
しかし、自分の環境では10回中9回はエディターが落ちてしまっている状況です。
こちらのSDKの水槽スペックに足りていないということでしょうか？
私のPCスペックは、CPU:13400F、GPU:3060ti、Ram:32GB、という構成になっております。
もし他の要因が考えられる場合は教えていただけますと幸いです。
ログファイルを添付いたしましたので何卒宜しくお願い致します。
[SimpleCollabSystem.log](https://github.com/user-attachments/files/18280003/SimpleCollabSystem.log)

**Reply by PaperSloth** [reply]
追加でログをいただきありがとうございます！
ログを見ていて気になった点として、他にもいくつかプラグインを使用されているでしょうか？
以下のログが気になりました👀
`LogPluginManager: Error: モジュール 'UEOSC' が見つからないためプラグイン 'UEOSC' のロードに失敗しました。プラグインが正しくインストールされていることを確認するか、このプロジェクトでは対象プラグインを無効にしてください。`
